File: src/cpp/src/stages/break_bond.cpp
#include "molgr/stages/break_bond.h"

#include "molgr/utils/electrons.h"
#include "molgr/utils/logger.h"
#include "molgr/utils/smarts.h"

#include <openbabel/atom.h>
#include <openbabel/bond.h>
#include "molgr/compat/openbabel_iter.h"

#include <algorithm>
#include <cmath>
#include <vector>

namespace molgr
{
    namespace reconstruct
    {
        using namespace OpenBabel;

        namespace
        {
            int CurrentBondBreakingElectronBudget(OBMol &mol)
            {
                int sum = 0;
                FOR_ATOMS_OF_MOL(atom, mol)
                {
                    sum += molgr::utils::GetUnpairedElectronCount(*atom);
                    if (molgr::utils::HasUnresolvedTwoElectronCenter(*atom))
                    {
                        sum += 2;
                    }
                }
                return sum;
            }
        }

        // Electron bookkeeping: reduce one deformed multiple bond order and add
        // one real unpaired electron to each endpoint (homolytic pi-bond cleavage).
        // Deferred two-electron centers count toward the stopping budget so their
        // occupied electrons are not created again by another bond cleavage.
        bool BreakDeformedEne(OBMol &mol, int allowed_charge, int allowed_radical, double tolerance)
        {
            bool hit = false;

            struct EneCandidate
            {
                int idx1;
                int idx2;
                double deviation;
            };
            std::vector<EneCandidate> candidates;

            {
                auto matches = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::BREAK_DEFORMED_ENE_A);
                for (const auto &idxs : matches)
                {
                    OBBond *bond = mol.GetBond(idxs[1], idxs[2]);
                    if (!bond || bond->IsRotor() || bond->GetBondOrder() == 1)
                        continue;

                    double torsion = std::abs(mol.GetTorsion(idxs[0], idxs[1], idxs[2], idxs[3]));
                    double deviation = std::min(torsion, 180.0 - torsion);
                    if (deviation > tolerance)
                    {
                        candidates.push_back(EneCandidate{idxs[1], idxs[2], deviation});
                    }
                }
            }

            {
                auto matches = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::BREAK_DEFORMED_ENE_B);
                for (const auto &idxs : matches)
                {
                    OBBond *bond = mol.GetBond(idxs[1], idxs[2]);
                    if (!bond || bond->IsRotor() || bond->GetBondOrder() == 1)
                        continue;

                    double torsion = std::abs(mol.GetTorsion(idxs[0], idxs[1], idxs[2], idxs[3]));
                    double deviation = std::min(torsion, 180.0 - torsion);
                    if (deviation > tolerance)
                    {
                        candidates.push_back(EneCandidate{idxs[1], idxs[2], deviation});
                    }
                }
            }

            std::sort(candidates.begin(), candidates.end(), [](const EneCandidate &a, const EneCandidate &b)
                      { return a.deviation > b.deviation; });

            for (const auto &candidate : candidates)
            {
                if (CurrentBondBreakingElectronBudget(mol) >=
                    std::abs(allowed_charge) + allowed_radical)
                    return hit;
                OBBond *bond = mol.GetBond(candidate.idx1, candidate.idx2);
                if (!bond || bond->IsRotor() || bond->GetBondOrder() == 1)
                    continue;
                bond->SetBondOrder(bond->GetBondOrder() - 1);
                OBAtom *begin_atom = bond->GetBeginAtom();
                OBAtom *end_atom = bond->GetEndAtom();
                molgr::utils::SetUnpairedElectronCount(*begin_atom, molgr::utils::GetUnpairedElectronCount(*begin_atom) + 1);
                molgr::utils::SetUnpairedElectronCount(*end_atom, molgr::utils::GetUnpairedElectronCount(*end_atom) + 1);
                hit = true;
                LOG_DEBUG("[BreakDeformedEne] Broken sorted candidate");
            }
            return hit;
        }

        // Electron bookkeeping: ordinary reductions/deletions create one unpaired
        // electron per endpoint. The cation template creates one radical only on
        // the neutral endpoint and lowers the positive endpoint charge, an explicit
        // charge-transfer heuristic. Deferred two-electron centers count toward
        // the stopping budget but remain unresolved and are not consumed here.
        bool BreakOneBond(OBMol &mol, int &charge, int allowed_radical)
        {
            bool hit = false;
            auto check_cond = [&]()
            {
                return CurrentBondBreakingElectronBudget(mol) >=
                       std::abs(charge) + allowed_radical;
            };

            while (true)
            {
                auto matches = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::BREAK_ONE_BOND_MULTIPLE);
                if (matches.empty())
                    break;
                if (check_cond())
                    return hit;
                const auto &idxs = matches.front();
                OBBond *bond = mol.GetBond(idxs[0], idxs[1]);
                if (!bond)
                    break;
                bond->SetBondOrder(bond->GetBondOrder() - 1);
                OBAtom *begin_atom = bond->GetBeginAtom();
                OBAtom *end_atom = bond->GetEndAtom();
                molgr::utils::SetUnpairedElectronCount(*begin_atom, molgr::utils::GetUnpairedElectronCount(*begin_atom) + 1);
                molgr::utils::SetUnpairedElectronCount(*end_atom, molgr::utils::GetUnpairedElectronCount(*end_atom) + 1);
                hit = true;
                LOG_DEBUG("[BreakOneBond] Broken triple/double candidate");
            }

            while (true)
            {
                auto matches = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::BREAK_ONE_BOND_CATION);
                if (matches.empty())
                    break;
                if (check_cond())
                    return hit;
                const auto &idxs = matches.front();
                OBBond *bond = mol.GetBond(idxs[0], idxs[1]);
                if (!bond)
                    break;
                bond->SetBondOrder(bond->GetBondOrder() - 1);
                OBAtom *idx0_atom = mol.GetAtom(idxs[0]);
                OBAtom *idx1_atom = mol.GetAtom(idxs[1]);
                if (!idx0_atom || !idx1_atom)
                    break;
                molgr::utils::SetUnpairedElectronCount(*idx1_atom, molgr::utils::GetUnpairedElectronCount(*idx1_atom) + 1);
                idx0_atom->SetFormalCharge(idx0_atom->GetFormalCharge() - 1);
                charge += 1;
                hit = true;
                LOG_DEBUG("[BreakOneBond] Broken charge-transfer candidate");
            }

            {
                auto matches = molgr::smarts::FindAll(mol, molgr::smarts::PatternId::BREAK_ONE_BOND_AROMATIC);
                if (!matches.empty())
                {
                    if (check_cond())
                        return hit;
                    for (const auto &idxs : matches)
                    {
                        OBBond *bond = mol.GetBond(idxs[0], idxs[1]);
                        if (!bond)
                            continue;
                        if (bond->GetBondOrder() == 1)
                            continue;
                        bond->SetBondOrder(bond->GetBondOrder() - 1);
                        OBAtom *begin_atom = bond->GetBeginAtom();
                        OBAtom *end_atom = bond->GetEndAtom();
                        molgr::utils::SetUnpairedElectronCount(*begin_atom, molgr::utils::GetUnpairedElectronCount(*begin_atom) + 1);
                        molgr::utils::SetUnpairedElectronCount(*end_atom, molgr::utils::GetUnpairedElectronCount(*end_atom) + 1);
                        hit = true;
                        LOG_DEBUG("[BreakOneBond] Broken aromatic candidate");
                        break;
                    }
                }
            }

            if (check_cond())
                return hit;

            bool all_single = true;
            FOR_BONDS_OF_MOL(b, mol)
            if (b->GetBondOrder() != 1) all_single = false;

            if (all_single)
            {
                std::vector<OBBond *> single_bonds;
                FOR_BONDS_OF_MOL(b, mol)
                {
                    single_bonds.push_back(&*b);
                }
                if (!single_bonds.empty())
                {
                    if (check_cond())
                        return hit;
                    OBBond *single_bond = single_bonds.front();
                    OBAtom *b_at = single_bond->GetBeginAtom();
                    OBAtom *e_at = single_bond->GetEndAtom();
                    molgr::utils::SetUnpairedElectronCount(*b_at, molgr::utils::GetUnpairedElectronCount(*b_at) + 1);
                    molgr::utils::SetUnpairedElectronCount(*e_at, molgr::utils::GetUnpairedElectronCount(*e_at) + 1);
                    mol.DeleteBond(single_bond);
                    hit = true;
                    LOG_DEBUG("[BreakOneBond] Deleted single bond");
                }
            }
            return hit;
        }
    }
}

File: src/molgr/fallback/stages/break_bond.py
"""Bond-breaking heuristics used when the deterministic cleanup path still stalls."""

from __future__ import annotations

from typing import List, Tuple, cast

from openbabel import openbabel as ob
from openbabel import pybel

from molgr.fallback.utils import smarts
from molgr.fallback.utils.electrons import (
    get_unpaired_electron_count,
    has_unresolved_two_electron_center,
    set_unpaired_electron_count,
)


def _bond_breaking_electron_budget(obmol: ob.OBMol) -> int:
    """Count electrons already represented before a recovery bond cleavage."""

    return sum(
        get_unpaired_electron_count(atom) + (2 if has_unresolved_two_electron_center(atom) else 0)
        for atom in ob.OBMolAtomIter(obmol)
    )


def break_deformed_ene(
    omol: pybel.Molecule,
    given_charge: int = 0,
    given_radical: int = 0,
    tolerance: float = 10,
) -> tuple[pybel.Molecule, bool]:
    """Homolytically reduce deformed multiple bonds by one bond-order unit.

    Each reduction creates one real unpaired electron on each endpoint. Deferred
    two-electron centers count toward the stopping budget so their electrons are
    not created again by another bond cleavage; their markers remain unresolved.
    """

    possible_ene_pairs: List[Tuple[int, int, float]] = []
    obmol = cast(ob.OBMol, omol.OBMol)
    matches: List[Tuple[int, int, int, int]] = list(smarts.BREAK_DEFORMED_ENE_A.findall(omol))
    while len(matches):
        idxs = matches.pop(0)
        bond2_1 = cast(ob.OBBond, obmol.GetBond(idxs[1], idxs[2]))
        if bond2_1.IsRotor() or bond2_1.GetBondOrder() == 1:
            continue
        torsion_angle: float = abs(obmol.GetTorsion(*idxs))
        torsion_angle = min(torsion_angle, 180 - torsion_angle)
        if torsion_angle > tolerance:
            possible_ene_pairs.append((idxs[1], idxs[2], torsion_angle))

    matches = list(smarts.BREAK_DEFORMED_ENE_B.findall(omol))
    while len(matches):
        idxs = matches.pop(0)
        bond2_2 = cast(ob.OBBond, obmol.GetBond(idxs[1], idxs[2]))
        if bond2_2.IsRotor() or bond2_2.GetBondOrder() == 1:
            continue
        torsion_angle = abs(obmol.GetTorsion(*idxs))
        torsion_angle = min(torsion_angle, 180 - torsion_angle)
        if torsion_angle > tolerance:
            possible_ene_pairs.append((idxs[1], idxs[2], torsion_angle))

    possible_ene_pairs.sort(key=lambda x: x[2], reverse=True)
    hit = False
    for idx1, idx2, _ in possible_ene_pairs:
        if _bond_breaking_electron_budget(obmol) >= abs(given_charge) + given_radical:
            return omol, hit
        bond = cast(ob.OBBond, obmol.GetBond(idx1, idx2))
        if bond.IsRotor() or bond.GetBondOrder() == 1:
            continue
        bond.SetBondOrder(bond.GetBondOrder() - 1)
        begin_atom = cast(ob.OBAtom, bond.GetBeginAtom())
        end_atom = cast(ob.OBAtom, bond.GetEndAtom())
        set_unpaired_electron_count(begin_atom, get_unpaired_electron_count(begin_atom) + 1)
        set_unpaired_electron_count(end_atom, get_unpaired_electron_count(end_atom) + 1)
        hit = True
    return omol, hit


def break_one_bond(
    omol: pybel.Molecule,
    given_charge: int = 0,
    given_radical: int = 0,
) -> tuple[pybel.Molecule, int, bool]:
    """Apply last-resort homolytic and charge-transfer bond-breaking templates.

    Multiple/aromatic bond reduction and single-bond deletion create one real
    unpaired electron at each endpoint. The cation template instead creates one
    radical only at the neutral endpoint while lowering the positive endpoint's
    formal charge, an explicit heterolytic/one-electron heuristic. Deferred
    two-electron centers count toward the stopping budget without being classified
    or consumed; active lone pairs do not count toward that budget.
    """

    obmol = cast(ob.OBMol, omol.OBMol)
    hit = False
    while res := smarts.BREAK_ONE_BOND_MULTIPLE.findall(omol):
        if _bond_breaking_electron_budget(obmol) >= abs(given_charge) + given_radical:
            return omol, given_charge, hit
        idxs = res.pop(0)
        bond = cast(ob.OBBond, obmol.GetBond(idxs[0], idxs[1]))
        bond.SetBondOrder(bond.GetBondOrder() - 1)
        begin_atom = cast(ob.OBAtom, bond.GetBeginAtom())
        end_atom = cast(ob.OBAtom, bond.GetEndAtom())
        set_unpaired_electron_count(begin_atom, get_unpaired_electron_count(begin_atom) + 1)
        set_unpaired_electron_count(end_atom, get_unpaired_electron_count(end_atom) + 1)
        hit = True

    while res := smarts.BREAK_ONE_BOND_CATION.findall(omol):
        if _bond_breaking_electron_budget(obmol) >= abs(given_charge) + given_radical:
            return omol, given_charge, hit
        idxs = res.pop(0)
        bond2 = cast(ob.OBBond, obmol.GetBond(idxs[0], idxs[1]))
        bond2.SetBondOrder(bond2.GetBondOrder() - 1)
        begin_atom2 = cast(ob.OBAtom, bond2.GetBeginAtom())
        end_atom2 = cast(ob.OBAtom, bond2.GetEndAtom())
        set_unpaired_electron_count(end_atom2, get_unpaired_electron_count(end_atom2) + 1)
        begin_atom2.SetFormalCharge(int(begin_atom2.GetFormalCharge() - 1))
        given_charge += 1
        hit = True

    res = list(smarts.BREAK_ONE_BOND_AROMATIC.findall(omol))
    if res:
        if _bond_breaking_electron_budget(obmol) >= abs(given_charge) + given_radical:
            return omol, given_charge, hit
        for idxs in res:
            bond3 = cast(ob.OBBond, obmol.GetBond(idxs[0], idxs[1]))
            if bond3.GetBondOrder() == 1:
                continue
            bond3.SetBondOrder(bond3.GetBondOrder() - 1)
            begin_atom3 = cast(ob.OBAtom, bond3.GetBeginAtom())
            end_atom3 = cast(ob.OBAtom, bond3.GetEndAtom())
            set_unpaired_electron_count(begin_atom3, get_unpaired_electron_count(begin_atom3) + 1)
            set_unpaired_electron_count(end_atom3, get_unpaired_electron_count(end_atom3) + 1)
            hit = True
            break

    if all(cast(ob.OBBond, bond).GetBondOrder() == 1 for bond in ob.OBMolBondIter(obmol)):
        single_bonds = list(ob.OBMolBondIter(obmol))
        if single_bonds:
            if _bond_breaking_electron_budget(obmol) >= abs(given_charge) + given_radical:
                return omol, given_charge, hit
            single_bond = cast(ob.OBBond, single_bonds[0])
            single_begin_atom = cast(ob.OBAtom, single_bond.GetBeginAtom())
            single_end_atom = cast(ob.OBAtom, single_bond.GetEndAtom())
            set_unpaired_electron_count(
                single_begin_atom, get_unpaired_electron_count(single_begin_atom) + 1
            )
            set_unpaired_electron_count(
                single_end_atom, get_unpaired_electron_count(single_end_atom) + 1
            )
            obmol.DeleteBond(single_bond)
            hit = True
    return omol, given_charge, hit


__all__ = ["break_deformed_ene", "break_one_bond"]
